### PR TITLE
Res 1179 customresnet

### PR DESCRIPTION
**Pull request by lucasosouza**
Customizable Resnets 1.5 with sparse weights and kwinners.  Allow customization per groups, which might be the most straightforward way to search hyperparameters, or full customization. 

Using named tuples since the attributes can be set and retrieved by name and have defaults, while still being exported as regular tuples. Two different named tuples:
- `LayerParams`: for layers with activation. The parameters in this tuple are the kwinners parameters plus weights_density (see below). 
- `NoactLayerParams`: for layers without activation. A single parameter is used, weights_density.

Usage example, customization per group:

```
    custom_sparse_params = dict(
        stem=LayerParams(),
        filters64=dict(
            conv1x1_1=LayerParams(
                percent_on=0.3,
                boost_strength=1.2,
                boost_strength_factor=1.0,
                local=False,
                weights_density=0.3,
            ),
            conv3x3_2=LayerParams(
                percent_on=0.1,
                boost_strength=1.2,
                boost_strength_factor=1.0,
                local=True,
                weights_density=0.1,
            ),
            conv1x1_3=NoactLayerParams(weights_density=0.1),
            shortcut=LayerParams(
                percent_on=0.4,
                weights_density=0.4,
            ),
        ),
        filters128=dict(
            conv1x1_1=LayerParams(),
            conv3x3_2=LayerParams(),
            conv1x1_3=NoactLayerParams(),
            shortcut=LayerParams(),
        ),
        filters256=dict(
            conv1x1_1=LayerParams(),
            conv3x3_2=LayerParams(),
            conv1x1_3=NoactLayerParams(),
            shortcut=LayerParams(),
        ),
        filters512=dict(
            conv1x1_1=LayerParams(),
            conv3x3_2=LayerParams(),
            conv1x1_3=NoactLayerParams(),
            shortcut=LayerParams(),
        ),
        linear=NoactLayerParams(weights_density=0.5),
    )
```

Each block can also be customized independently. In this case, the value of a group key (such as `filters64`) would be a list with dictionaries equivalent to the number of blocks in the group, as opposed to a single dictionary. In ResNet50 that multiplies the number of hyperparameters to be optimized by 16 , since the group sizes in ResNet 50 are `[3,4,6,3]`, summing up to 16 blocks. 

Usage example, full customization:
```
        filters64=[  # 3 blocks
            dict(
                conv1x1_1=LayerParams(
                    percent_on=0.3,
                    boost_strength=1.2,
                    boost_strength_factor=1.0,
                    local=False,
                    weights_density=0.3,
                ),
                conv3x3_2=LayerParams(),
                conv1x1_3=NoactLayerParams(),
                shortcut=LayerParams(),
            ),
            dict(
                conv1x1_1=LayerParams(),
                conv3x3_2=LayerParams(),
                conv1x1_3=NoactLayerParams(),
                shortcut=LayerParams(),
            ),
            dict(
                conv1x1_1=LayerParams(),
                conv3x3_2=LayerParams(),
                conv1x1_3=NoactLayerParams(),
                shortcut=LayerParams(),
            ),
        ],
```

Network has 2 extra parameters to define for the entire network the type of sparse layer to be used:
```
            linear_sparse_weights_type="SparseWeights",
            conv_sparse_weights_type="SparseWeights2d",
```

------------------------

Current top-1 accuracy after 50 epochs running:
- using defaults for `LayerParams` and `NoactLayerParams`
- Adam optimizer with lr 2e-3, no weight decay

image

### CODE REVIEW

File: nupic/research/frameworks/pytorch/models/resnets.py [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+def resnet101(config=None):
+    return build_resnet(101, config)
+
+
+def resnet152(config=None):
+    return build_resnet(152, config)
+
+
+# base tests
+if __name__ == "__main__":
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/models/resnets.py]: These should be in tests/ directory as unit tests.

**Comment by subutai** [comment]
> It doesn't pass the tests because the PR uses a feature from python 3.7 (`defaults` in named tuples). Are we still using python 3.6? If yes, is that a hard requirement?

That's a question for @lscheinkman !

File: nupic/research/frameworks/pytorch/models/resnets.py [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+    https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py"""
+
+    def __init__(self, config=None):
+        super(ResNet, self).__init__()
+
+        # update config
+        defaults = dict(
+            depth=50,
+            num_classes=10,
+            sparse_linear_layer_type="SparseWeights",
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/models/resnets.py]: Should have `weights` in the name, e.g. `linear_sparse_weights_type`

File: nupic/research/frameworks/pytorch/models/resnets.py [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+            local=local
+        )
+
+
+class BasicBlock(nn.Module):
+    """Default block for ResNets with < 50 layers."""
+
+    expansion = 1
+
+    def __init__(self, in_planes, planes, sparse_layer_type, layer_params, stride=1):
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/models/resnets.py]: `sparse_weights_type`

File: nupic/research/frameworks/pytorch/models/resnets.py [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+        out = self.post_activation(out)
+        return out
+
+
+class Bottleneck(nn.Module):
+    """Default block for ResNets with >= 50 layers."""
+
+    expansion = 4
+
+    def __init__(self, in_planes, planes, sparse_layer_type, layer_params, stride=1):
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/models/resnets.py]: `sparse_weights_type`

**Comment by lscheinkman** [comment]
@lucasosouza we could upgrade nupic.research python to 3.7.
I would create a separate PR to keep potential 3.6 to 3.7 problems isolated from this PR.

You will need to change the following files:
- environment.yml
- setup.py
- .circleci/config.yml

We don't regression tests so you may need to validate some of the experiments manually
